server/src/main/java/org/apache/iotdb/db/metadata/MManager.java

```diff
@@ -1846,4 +1836,219 @@ public void createMTreeSnapshot() {
+            dataType,
+            getDefaultEncoding(dataType),
+            TSFileDescriptor.getInstance().getConfig().getCompressor(),
+            Collections.emptyMap());
+        }
+
+        MeasurementMNode measurementNode = (MeasurementMNode) getChild(deviceNode, measurementList[i]);
+
+        // check type is match
+        TSDataType insertDataType = null;
+        if (plan instanceof InsertPlan) {
+          if (!((InsertPlan)plan).isNeedInferType()) {
+            // only when InsertPlan's values is object[], we should check type
+            insertDataType = getTypeInLoc(plan, i);
+          } else {
+            insertDataType = measurementNode.getSchema().getType();
+          }
+        } else if (plan instanceof InsertTabletPlan) {
+          insertDataType = getTypeInLoc(plan, i);
+        }
+
+        if (measurementNode.getSchema().getType() != insertDataType) {
+          logger.warn("DataType mismatch, Insert measurement {} type {}, metadata tree type {}",
+            measurementList[i], insertDataType, measurementNode.getSchema().getType());
+          if (!config.isEnablePartialInsert()) {
+            throw new MetadataException(String.format(
+              "DataType mismatch, Insert measurement %s type %s, metadata tree type %s",
+              measurementList[i], insertDataType, measurementNode.getSchema().getType()));
+          } else {
+            // mark failed measurement
+            if (plan instanceof InsertTabletPlan) {
+              ((InsertTabletPlan) plan).markMeasurementInsertionFailed(i);
+            } else if (plan instanceof InsertPlan) {
+              ((InsertPlan) plan).markMeasurementInsertionFailed(i);
+            }
+            continue;
+          }
+        }
+
+        // maybe need to convert value type to the true type
+        if ((plan instanceof InsertPlan)) {
+          changeValueToRealType((InsertPlan) plan, i, measurementNode.getSchema().getType());
+        }
+
+        schemas[i] = measurementNode.getSchema();
+        if (schemas[i] != null) {
+          measurementList[i] = schemas[i].getMeasurementId();
+        }
+      } catch (MetadataException e) {
+        logger.warn("meet error when check {}.{}, message: {}", deviceId, measurementList[i],
+          e.getMessage());
+        if (config.isEnablePartialInsert()) {
+          // mark failed measurement
+          if (plan instanceof InsertPlan) {
+            ((InsertPlan) plan).markMeasurementInsertionFailed(i);
+          } else if (plan instanceof InsertTabletPlan) {
+            ((InsertTabletPlan) plan).markMeasurementInsertionFailed(i);
+          }
+        } else {
+          throw e;
+        }
+      }
+    }
+    return schemas;
+  }
+
+  private void changeValueToRealType(InsertPlan plan, int loc, TSDataType type) throws MetadataException {
+    plan.getTypes()[loc] = type;
+    try {
+      switch (type) {
+        case INT32:
+          if (!(plan.getValues()[loc] instanceof Integer)) {
```

Comment:
Do we need this check? I think only String value need to be inferred.

Reply:
yes, could remove this check.